src/EventSubscriber/UserSoftDeleteSubscriber.php

```diff
@@ -0,0 +1,62 @@
+<?php
+
+namespace App\EventSubscriber;
+
+use App\Event\TrainerSoftDeleteEvent;
+use App\Event\UserSoftDeleteEvent;
+use App\Repository\ProgrammeRepository;
+use App\Repository\UserRepository;
+use Doctrine\ORM\OptimisticLockException;
+use Doctrine\ORM\ORMException;
+use Psr\EventDispatcher\EventDispatcherInterface;
+use Psr\Log\LoggerAwareTrait;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+class UserSoftDeleteSubscriber implements EventSubscriberInterface
+{
+    use LoggerAwareTrait;
+
+    private EventDispatcherInterface $eventDispatcher;
+
+    private UserRepository $userRepository;
+
+    private ProgrammeRepository $programmeRepository;
+
+    public function __construct(
+        EventDispatcherInterface $eventDispatcher,
+        UserRepository $userRepository,
+        ProgrammeRepository $programmeRepository
+    ) {
+        $this->eventDispatcher = $eventDispatcher;
+        $this->userRepository = $userRepository;
+        $this->programmeRepository = $programmeRepository;
+    }
+
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            UserSoftDeleteEvent::NAME => [
+                ['checkIfTrainer', 10],
+                ['softDeleteUser', 9],
+            ],
+        ];
+    }
+
+    public function checkIfTrainer(UserSoftDeleteEvent $event): void
```

Comment:
this is the only method that should actually be in a listener. also, search for `Gedmo\SoftDeleteable\SoftDeleteableListener` class and see the events that are dispatched there. You don't really need your own events